plugins/catalog-backend/src/ingestion/processors/GithubReaderProcessor.ts

```diff
@@ -15,50 +15,231 @@
-      config?.getOptionalString('catalog.processors.github.privateToken') ?? '';
+export function getApiRequestOptions(provider: ProviderConfig): RequestInit {
+  const headers: HeadersInit = {
+    Accept: 'application/vnd.github.v3.raw',
+  };
+
+  if (provider.token) {
+    headers.Authorization = `token ${provider.token}`;
   }
 
-  getRequestOptions(): RequestInit {
-    const headers: HeadersInit = {
-      Accept: 'application/vnd.github.v3.raw',
-    };
+  return {
+    headers,
+  };
+}
+
+export function getRawRequestOptions(provider: ProviderConfig): RequestInit {
+  const headers: HeadersInit = {};
+
+  if (provider.token) {
+    headers.Authorization = `token ${provider.token}`;
+  }
+
+  return {
+    headers,
+  };
+}
+
+// Converts for example
+// from: https://github.com/a/b/blob/branchname/path/to/c.yaml
+// to:   https://api.github.com/repos/a/b/contents/path/to/c.yaml?ref=branchname
+export function getApiUrl(target: string, provider: ProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+
+    if (
+      !owner ||
+      !name ||
+      !ref ||
+      (filepathtype !== 'blob' && filepathtype !== 'raw') ||
+      !filepath?.match(/\.ya?ml$/)
+    ) {
+      throw new Error('Wrong URL or invalid file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+    return new URL(
+      `${provider.apiBaseUrl}/repos/${owner}/${name}/contents/${pathWithoutSlash}?ref=${ref}`,
+    );
+  } catch (e) {
+    throw new Error(`Incorrect URL: ${target}, ${e}`);
+  }
+}
+
+// Converts for example
+// from: https://github.com/a/b/blob/branchname/c.yaml
+// to:   https://raw.githubusercontent.com/a/b/branchname/c.yaml
+export function getRawUrl(target: string, provider: ProviderConfig): URL {
+  try {
+    const { owner, name, ref, filepathtype, filepath } = parseGitUri(target);
+
+    if (
+      !owner ||
+      !name ||
+      !ref ||
+      (filepathtype !== 'blob' && filepathtype !== 'raw') ||
+      !filepath?.match(/\.ya?ml$/)
+    ) {
+      throw new Error('Wrong URL or invalid file path');
+    }
+
+    const pathWithoutSlash = filepath.replace(/^\//, '');
+    return new URL(
+      `${provider.rawBaseUrl}/${owner}/${name}/${ref}/${pathWithoutSlash}`,
+    );
+  } catch (e) {
+    throw new Error(`Incorrect URL: ${target}, ${e}`);
+  }
+}
+
+export function readConfig(config: Config): ProviderConfig[] {
+  const providers: ProviderConfig[] = [];
 
-    if (this.privateToken !== '') {
-      headers.Authorization = `token ${this.privateToken}`;
+  // In a previous version of the configuration, we only supported github,
+  // and the "privateToken" key held the token to use for it. The new
+  // configuration method is to use the "providers" key instead.
+  const providerConfigs =
+    config.getOptionalConfigArray('catalog.processors.github.providers') ??
+    config.getOptionalConfigArray('catalog.processors.githubApi.providers') ??
```

Comment:
Could we just throw an error explaining that it's been moved to `*.github.*` if `catalog.processors.githubApi.providers` exists? Makes it easier to remove in the future. Same with the ones below

Reply:
Sure, because alpha? It's a bit jarring for those who already have it configured.

Reply:
Warning as @blam says is a middle ground but I'm not sure who reads the log output carefully enough :)

Reply:
Yeah, I'd say the non-alpha thing to do would be a deprecation warning, which we could do, and then later move to an error.

It might be nicer in this case seeing as we're expecting people to be using the config that we're breaking. As long as we're providing a clear enough fix I don't think we'll waste anyone's time too much